exonum/src/messages/mod.rs

```diff
@@ -54,6 +56,7 @@ mod tests;
 
 /// Version of the protocol. Different versions are incompatible.
 pub const PROTOCOL_MAJOR_VERSION: u8 = 1;
+pub(crate) const RAW_TRANSACTION_HEADER: usize = mem::size_of::<u16>() * 2;
 
 /// Transaction raw buffer.
 /// This struct is used to transfer transactions in network.
```

Comment:
You mean that user can define his own order, and we shouldn't order transaction in lexicographic order?

But `Eq` probably still shoud be present?

Reply:
> You mean that user can define his own order, and we shouldn't order transaction in lexicographic order?

Yeah, I see no compelling reason to derive ordering traits if they are not used internally. OTOH, ordering traits must agree with `PartialEq` implementation, so there is an argument for deriving them (or at least documenting how equivalence is implemented for transactions).

> But `Eq` probably still shoud be present?

Not 100% sure about this. I had a mistaken impression that `Eq` assumes completeness whereas `PartialEq` doesn't, but this isn't the case.

Reply:
> I see no compelling reason to derive ordering traits if they are not used internally

But `RawTransaction` is part of public interface. So deriving `PartialEq`\`Ord` allows user to put `RawTransaction` into map

Reply:
For public types it's better to implement standard traits if there is a reasonable expected implementation. Otherwise users will not be happy about newtype wrappers they have to implement and use in order to do something that requires standard traits (like storing values in map collections, using `assert_eq!`, etc.)

Reply:
> For public types it's better to implement standard traits if there is a reasonable expected implementation.

Generally, that makes sense, but to a certain point. If semantically there is no complete ordering among transactions, implementing ordering traits for them may not make much sense. (Like, what does `Transfer { to: alice, amount: 10, .. } < Transfer { to: bob, amount: 5, .. }` statement mean semantically?)

> storing values in map collections

You can use `HashMap` for that (i.e., derive `Hash`); it's more obvious that a transaction can be hashed (we actually do that, albeit with a cryptographic hash function) than defining an order on them.

> using `assert_eq!`

Requires `PartialEq`.

Reply:
`HashMap` requieres `Eq` as well https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.remove

> Generally, that makes sense, but to a certain point. If semantically there is no complete ordering among transactions, implementing ordering traits for them may not make much sense. (Like, what does Transfer { to: alice, amount: 10, .. } < Transfer { to: bob, amount: 5, .. } statement mean semantically?)

Semantically `RawTransaction` is just a bunch of bytes, so i am sure it's ok to sort it lexicographically.

Reply:
> Semantically RawTransaction is just a bunch of bytes, so i am sure it's ok to sort it lexicographically.

You're right, it is _possible_ to order them lexicographically, but it does not mean that it _makes sense_ to do so. It's like ordering JPEG images lexicographically by their on-disk representation. It is a sound ordering, but it has little value other than having _some_ ordering as such. It may be useful for storing transactions in a binary sort tree, but it has no other uses.

Practically speaking, there may be multiple useful orderings of transactions (by author, by date, by size, by content), some of which are only partial (i.e., not complete, not suitable for `Ord`). There is no self-evident, natural, or unique ordering of transactions which is going to be used most of the time, if not always.

Thus I support @slowli in that we don't need `PartialOrd` or `Ord` on transactions.